text/unification.md

```diff
@@ -0,0 +1,268 @@
+
+When defining an existential variable with a term, the term is retyped. Is this really needed?
+
+# Directions for short terms improvements
+
+All the numbered points above are worth being addressed.
+
+My feeling is that the true backtracking ability C.1 of `Evarconv.solve_unif_constraints_with_heuristics` should come first followed by the extensions C.3 then C.2, so that the heuristics are strong enough to absorb/compensate incremental modifications in `evar_conv_x` and `solve_simple_eqn`.
+
+# Miscellaneous further perpectives
+
+**Delaying choices**
+- A way should be found to postpone the use of first-order unification heuristics as well as to postpone the decision of using canonical structures.
+  - For instance, for a `?P ?e =?= Q t` problem, one can set `?P := fun x => ?P'` with pending equation `?P'@{x:=?e} =?= Q t` and one could declare that `?e` has by default to be instantiated by `t` if no other instances is available.
+  - Similarly, for a `projᵢ ?e =?= t` problem, `?e` could be resolved into `Build ?e₁ ... t ... ?eₙ`, for `Build` the constructor of the record type, with some indication that each such subterm `Build ?e₁ ... t ... ?eₙ` with unsolved component should be resolved using a canonical equipment.
+
+**Extensions**
+- Extend canonical structures to canonical equipments based on knowing two or more projections?
+- Reorganize canonical structures so that it more generally solves equations of the form `proj₁ᵢ₁ (... projₙᵢₙ ?e ...) = u`. On one side, it should a priori give more discriminating solutions. On the other side, it would allow to treat canonical structures only in the flexible case.
+- Extend first-order unification to `match` term, e.g. to resolve `if true then t else u =?= if ?e₀ then ?e₁ else ?e₂` by setting `?e₀ := true`, `?e₁ := t` and `?e₂ := u` (see wish [#12780](https://github.com/coq/coq/issues/12780)).
+- Existential variables for sorts can sometimes not be instantiated by `Prop` even in situation where it would be legitimate to do so (see PR [#7369](https://github.com/coq/coq/pull/7369)).
+
+**Optimizations**
+- Systematically eta-reduce the solutions to existential variables (with all kinds of eta). On one side, this seems to be the natural expectation. On the other side, this would avoid caring about whether it might be dommageable to eta-expand this or that solution in the unification algorithm.
+- Give an order between constants which would allow to predict which side to reduce in `E[c] =?= E'[c]`.
+- Keep constants unfolding to fixpoints folded so as to concentrate on the essential and reduce the size of expressions, including in evaluation contexts (i.e. stacks).
+- Do more aggressive restrictions of contexts (e.g. Sozeau and Ziliani seem to take the downwards well-typed closure of restrictions while Coq takes the upwards well-typed closure - knowing that taking the closure by dependency requires in theory to take care of erasable occurrences)?
```

Comment:
What's the difference between downwards and upwards here?  (Or what's an example?)

Reply:
By `downwards` I mean the largest well-formed context which exclude the forbidden variables. By `upwards`, I mean the smallest well-formed context which include the allowed variables.

Reply:
Hmmm, it seems that upwards is more powerful unless we have reduced enough that no further occurrences of forbidden variables can be erased by reduction.  Consider `T : Type, f : T -> Type, x : T |- ?e : Type`, and the pair of problems `A, B : Type, f : bool -> Type, x : bool |- ?e@{T:=(fun _ => bool) A,f:=f,x:=x} == ?e@{T:=(fun _ => bool) B,f:=f,x:=x}` and `f : bool -> Type, x : bool |- ?e@{T:=bool,f:=f,x:=x} == f x`.  IIUC, upwards closure would keep `T`, while downwards would remove it?  (Do you have an example not involving reduction where they differ where upwards closure does better than downwards closure?)

Reply:
But maybe this is what you meant by "erasable"?